plugins/engine-datafusion/jni/src/lib.rs

```diff
@@ -338,26 +473,102 @@ impl CustomFileMeta {
+            error!("Runtime manager not initialized");
+            set_object_result_error(&mut env, callback,
+                                    &DataFusionError::Execution("Runtime manager not initialized".to_string()));
+            return;
+        }
+    };
+
+    // ===== EXTRACT ALL JAVA DATA BEFORE ASYNC BLOCK =====
+    let table_name: String = match env.get_string(&table_name) {
+        Ok(s) => s.into(),
+        Err(e) => {
+            error!("Failed to get table name: {}", e);
+            set_object_result_error(&mut env, callback,
+                                    &DataFusionError::Execution(format!("Failed to get table name: {}", e)));
+            return;
+        }
+    };
+
+    let plan_bytes_obj = unsafe { JByteArray::from_raw(substrait_bytes) };
+    let plan_bytes_vec = match env.convert_byte_array(plan_bytes_obj) {
+        Ok(bytes) => bytes,
+        Err(e) => {
+            error!("Failed to convert plan bytes: {}", e);
+            set_object_result_error(&mut env, callback,
+                                    &DataFusionError::Execution(format!("Failed to convert plan bytes: {}", e)));
+            return;
+        }
+    };
+
+    // Convert callback to GlobalRef (thread-safe)
+    let callback_ref = match env.new_global_ref(&callback) {
+        Ok(r) => r,
+        Err(e) => {
+            error!("Failed to create global ref: {}", e);
+            set_object_result_error(&mut env, callback,
+                                    &DataFusionError::Execution(format!("Failed to create global ref: {}", e)));
+            return;
+        }
+    };
+    let io_runtime = manager.io_runtime.clone();
+    let cpu_executor = manager.cpu_executor();
+
     let shard_view = unsafe { &*(shard_view_ptr as *const ShardView) };
     let runtime = unsafe { &*(runtime_ptr as *const DataFusionRuntime) };
-    let table_name: String = env
-        .get_string(&table_name)
-        .expect("Couldn't get java string!")
-        .into();
-
-    let runtimeEnv = &runtime.runtime_env;
 
     let table_path = shard_view.table_path();
     let files_meta = shard_view.files_metadata();
+
+    // Spawn async task - TRULY NON-BLOCKING!
+    io_runtime.block_on(async move {
+
+        let result = execute_query_with_cross_rt_stream(
+            table_path,
+            files_meta,
+            table_name,
+            plan_bytes_vec,
+            runtime,
+            cpu_executor,
+        ).await;
+
+        match result {
+            Ok(stream_ptr) => {
+                with_jni_env(|env| {
+                    set_object_result_ok_global(env, &callback_ref, stream_ptr as *mut u8);
+                });
+            }
+            Err(e) => {
+                with_jni_env(|env| {
+                    error!("Query execution failed: {}", e);
+                    set_object_result_error_global(env, &callback_ref, &e);
+                });
+            }
+        }
+    });
+}
+
+async fn execute_query_with_cross_rt_stream(
```

Comment:
Should we move all query execution to a separate file with the DF query execution logic while retaining the JNI logic in this file?